WP3/D3.8/report.tex

```diff
@@ -185,13 +185,94 @@ \section{Continuous integration achievements of OpenDreamKit projects}\label{pro
+test changes that involve dependent software packages. This can be frustrating
+as testing such changes is often the most time consuming task for developers
+and it would be particularly nice to give such tasks away to a CI system.
+
+\subsubsection{Modern CI with GitLab CI and CircleCI}
+To fix these issues we were looking at best pratices that would give the
+SageMath community fast, reproducible, reliable CI with minimal maintenance and
+continuous development necessary to keep the CI working for everybody.  We
+settled for GitLab CI, a modern open-source CI system, and CircleCI, a popular
+closed-source system. Which one to use is up to each developer. For simplicity
+we focus on GitLab CI here. GitLab CI typically uses docker containers as its
+backend (but it can also be based on many other technologies.) These containers
+are essentially virtual machines that can be hosted in a central location but
+also on private computers. Unlike in the patchbot systems, docker containers
+are standardized and isolated so that there is usually no issue of false
+positives. % note that I do not distinguish between docker containers and
+% images; I think it's more confusing to the reader to make such a distinction.
+
+GitLab CI now runs a "pipeline" for every change proposed by a developer.
+\TODO{Insert \url{https://trac.sagemath.org/raw-attachment/ticket/24655/gitlab.png} here.}
+This pipeline goes in stages. In the setup for SageMath, the first stage builds
+several flavors of docker container which contain a built version of Sage. The
+second stage runs tests on these containers. Finally the docker containers from
+the first stage are made publicly available.
+
+It should be noted that the first stage is where most complications for Sage
+arise. Building Sage takes several hours on most machines, so a lot of effort
+went into speeding this up in a sane and reproducible way to about 10 minutes.
+Also, we try to make the resulting docker containers as small as possible. This
+speeds up the whole CI pipeline and also makes it easier for developers to
+download these containers; in typical cases they only have to download a few MB
+to get a hold of the docker container.
+
+\subsubsection{Use Cases of Docker Containers}
+Having these docker containers readily available as a by-product of our CI turned out to be very valuable.
+\TODO{Maybe insert \url{https://trac.sagemath.org/raw-attachment/ticket/24655/dockerhub.png} here.}
+
+It means that recent stable \& unstable versions of Sage are automatically made
+available for anybody to work with, without the need of having to understand
+how to build Sage on their machine.
+
+Sage developers can use these to reproduce and understand issues
+reported by the CI system. Simply by downloading the container and running the
+Sage contained in that container.
+
+Projects that depend on Sage can use these docker containers for their own CI
```

Comment:
It might be worth noting explicitly here that CI systems have been moving largely towards using containerization technology in their build infrastructure, and thus giving users the ability to supply container images as the basis for their build, rather than relying solely on the default build images provided by the CI system.